testing/trino-testing/src/main/java/io/trino/testing/AbstractTestQueryFramework.java

```diff
@@ -83,18 +82,34 @@
 {
     private static final SqlParser SQL_PARSER = new SqlParser();
 
+    private boolean initialized;
     private QueryRunner queryRunner;
     private H2QueryRunner h2QueryRunner;
     private final AutoCloseableCloser afterClassCloser = AutoCloseableCloser.create();
     private io.trino.sql.query.QueryAssertions queryAssertions;
 
-    @BeforeClass
-    public void init()
-            throws Exception
+    // TODO (https://github.com/trinodb/trino/issues/11294) make this back as `@BeforeClass public void init()`
+    //   - remove all explicit callers
+    //   - remove synchronization
+    protected synchronized void initIfNeeded()
```

Comment:
feels like this should be `private` as subclasses cannot control the call-lifecycle of this method (cannot access `initialized` variable)

Reply:
it needs to be available to subclasses

- let's say `createQueryRunner` allocates some resources (eg docker container)
- and a subclass's test method uses that container (before using the query runner)

for this to work, the subclass needs to invoke `initIfNeeded` explicitly.

(yes, this is a breaking change)

Reply:
Ok so the assumption is that `initIfNeeded` would not be overridden - just called by the subclasses. And any subclass-specific initialization goes to `createQueryRunner` I guess the structure makes sense then.

Reply:
> Ok so the assumption is that `initIfNeeded` would not be overridden

yes! i made it `final` to convey that.